### PR TITLE
scm: implement gitignore

### PR DESCRIPTION
**Pull request by skshetry**
[**UPDATE**]: Went to implement `scm._reset()` function and use `dulwich` by caching its `IgnoreManager`. Regarding `pygit2`, leaving as-is till the migrations.

---

If we start making a lot of `is_ignored` checks, `dulwich` is a lot slower than `pygit2`. I'm toying with the idea of not walking through `.gitignore`-d files/directories for `repo.stage` and `repo.graph` collection.

On master version, the performance on `dvc status` is following:
```console

  Time (mean ± σ):      4.240 s ±  0.192 s    [User: 3.535 s, System: 0.474 s]
  Range (min … max):    4.141 s …  4.781 s    10 runs
```

On a modified version that does not traverse through git-ignored directory, the performance with the `dulwich`/`gitpython` and `pygit2` respectively are:
```console
Benchmark #1: dvc status with dulwich
  Time (mean ± σ):      2.260 s ±  0.389 s    [User: 1.779 s, System: 0.074 s]
  Range (min … max):    1.837 s …  2.883 s    10 runs

Benchmark #2: dvc status with gitpython
  Time (mean ± σ):      1.740 s ±  0.317 s    [User: 874.2 ms, System: 719.3 ms]
  Range (min … max):    1.485 s …  2.604 s    10 runs

Benchmark #3: dvc status with pygit2
  Time (mean ± σ):     461.9 ms ±   8.8 ms    [User: 337.5 ms, System: 47.3 ms]
  Range (min … max):   449.1 ms … 476.3 ms    10 runs
```

Note that `dulwich`'s performance can be made comparable to pygit2 if we cache the `IgnoreManager`, but by doing so, we won't be able to use it easily in the API (we need to `reset` it somewhere/somehow).
Just to compare, this is the performance with `IgnoreManager` cached.
```console
Benchmark #1: dvc status with dulwich cached
  Time (mean ± σ):     512.1 ms ±  73.2 ms    [User: 387.6 ms, System: 43.3 ms]
  Range (min … max):   436.2 ms … 704.0 ms    10 runs
```

There is deviation in `dulwich` for some reason though, but it comparable with pygit2.

* [x] ❗ I have followed the [Contributing to DVC](https://dvc.org/doc/user-guide/contributing/core) checklist.

* [x] 📖 If this PR requires [documentation](https://dvc.org/doc) updates, I have created a separate PR (or issue, at least) in [dvc.org](https://github.com/iterative/dvc.org) and linked it here.

Thank you for the contribution - we'll try to review it as soon as possible. 🙏

### CODE REVIEW

File: dvc/scm/git/backend/dulwich.py [hunk 1]
```diff
@@ -223,11 +225,14 @@ def _get_stash(self, ref: str):
...
+    @cached_property
+    def ignore_manager(self):
+        from dulwich.ignore import IgnoreFilterManager
+
+        return IgnoreFilterManager.from_repo(self.repo)
 
-        manager = ignore.IgnoreFilterManager.from_repo(self.repo)
-        return manager.is_ignored(relpath(path, self.root_dir))
+    def is_ignored(self, path: str) -> bool:
+        return self.ignore_manager.is_ignored(relpath(path, self.root_dir))
```
Comment (skshetry) [inline dvc/scm/git/backend/dulwich.py:235]: Looks like I have to drop this change and start using `pygit2`.

**Comment by skshetry** [comment]
Okay, decided to implement `_reset` anyway. :)

**Review by pmrowla — APPROVED** [review]

**Comment by efiop** [comment]
Great stuff, @skshetry , thank you! :pray:

**Comment by skshetry** [comment]
@efiop, the above benchmarks are with  #5244. This PR can be considered as a prerequisite for that issue, if we decide to move in with it

**Pull request merged**